### PR TITLE
Fix fit range for multiple inputs in indirect data analysis

### LINKED ISSUE #24664: Indirect Data Analysis - Energy range doesn't change when switching between workspaces
**Original reporter:** Spencer

### Actual behavior
The energy range in the mini-plots is not changing when switching between two workspaces with differing energy ranges.

### Steps to reproduce the behavior
1.`Interfaces`->`Indirect`->`Data Analysis`->`ConvFit` tab
2. Go to Multiple Input and load the data below
3. Click Run and wait.
4. Select the OSIRIS workspace using the combobox
5. The energy range does not change in the mini-plots! Notice that the EndX and StartX in the data table is now -0.9 to 2.1 (which is correct), but this is not reflected in the mini-plot

**Data Files**
[ConvFit Files.zip](https://github.com/mantidproject/mantid/files/2817435/ConvFit.Files.zip)

### Platforms affected
All

### PR DESCRIPTION
**Pull request by Harrietbrown**
**Description of work.**

This PR fixes an issue with the fitting of multiple spectra in indirect data analysis where it would only use the fit range of the first spectra.

**Report to:** Spencer

**To test:**

1.`Interfaces`->`Indirect`->`Data Analysis`->`ConvFit` tab
2. Go to Multiple Input and load the data below
3. Click Run and wait.
4. Select the OSIRIS workspace using the combobox
5. Note that the fit range of the OSIRIS is the same as that listed in the table.

**Data Files**
[ConvFit Files.zip](https://github.com/mantidproject/mantid/files/2817435/ConvFit.Files.zip)

Fixes #24664 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/CurveFitting/src/Algorithms/IqtFit.cpp [hunk 1]
```diff
@@ -164,19 +167,27 @@ double IqtFit<QENSFitSimultaneous>::getStartX(std::size_t index) const {
...
-template <typename Base>
-double IqtFit<Base>::getStartX(std::size_t /*unused*/) const {
-  return Base::getProperty("StartX");
+template <> double IqtFit<QENSFitSequential>::getStartX(std::size_t i) const {
+  std::vector<double> startX = QENSFitSequential::getProperty("StartX");
+  if (startX.size() == 1) {
+    return startX[0];
+  } else {
+    return startX[i];
+  }
```
Comment (robertapplin) [inline Framework/CurveFitting/src/Algorithms/IqtFit.cpp:176]: Should we check if `i >= startX.size()` ?

File: Framework/CurveFitting/src/Algorithms/IqtFit.cpp [hunk 2]
```diff
@@ -164,19 +167,27 @@ double IqtFit<QENSFitSimultaneous>::getStartX(std::size_t index) const {
...
-template <typename Base>
-double IqtFit<Base>::getEndX(std::size_t /*unused*/) const {
-  return Base::getProperty("EndX");
+template <> double IqtFit<QENSFitSequential>::getEndX(std::size_t i) const {
+  std::vector<double> endX = QENSFitSequential::getProperty("EndX");
+  if (endX.size() == 1) {
+    return endX[0];
+  } else {
+    return endX[i];
+  }
```
Comment (robertapplin) [inline Framework/CurveFitting/src/Algorithms/IqtFit.cpp:190]: Same here as above

File: Framework/CurveFitting/src/Algorithms/QENSFitSequential.cpp [hunk 3]
```diff
@@ -586,9 +598,22 @@ void QENSFitSequential::exec() {
...
-  addAdditionalLogs(resultWs);
-  copyLogs(std::dynamic_pointer_cast<MatrixWorkspace>(resultWs->getItem(0)),
-           groupWs);
+  size_t itter = 0;
+  for (auto results : resultWs->getAllItems()) {
+    addAdditionalLogs(results);
+    std::string resultWsName = results->getName();
+    auto endLoc = resultWsName.find("__Result");
+    std::string baseName = resultWsName.erase(endLoc);
+    for (auto workspace : groupWs->getAllItems()) {
```
Comment (robertapplin) [inline Framework/CurveFitting/src/Algorithms/QENSFitSequential.cpp:607]: Can `workspace` be a `&`? And the name strings be `const` ?

File: Framework/CurveFitting/test/Algorithms/QENSFitSequentialTest.h [hunk 4]
```diff
@@ -72,22 +96,48 @@ class QENSFitSequentialTest : public CxxTest::TestSuite {
...
     testFitOutput(outputBaseName, names.size() * 3);
     AnalysisDataService::Instance().clear();
   }
 
 private:
   std::string runConvolutionFit(const MatrixWorkspace_sptr &inputWorkspace,
-                                const MatrixWorkspace_sptr &resolution) {
+                                const MatrixWorkspace_sptr &resolution,
+                                const std::vector<double> startX,
+                                const std::vector<double> endX) {
```
Comment (robertapplin) [inline Framework/CurveFitting/test/Algorithms/QENSFitSequentialTest.h:133]: Can these vector parameters be `&` ?

**Review by robertapplin — CHANGES_REQUESTED** [review]
This fixes the previous problem with the fit ranges, Just a few minor suggestions

File: Framework/CurveFitting/src/Algorithms/IqtFit.cpp (hunk 1 above)
Comment (Harrietbrown) [reply Framework/CurveFitting/src/Algorithms/IqtFit.cpp:176]: this part of Iqtfit inherits from QNESSequential and uses it's validation, that means that startX must either contain a single value, or the same number of values as the size of the input workspace. So `i` can not be greater than startX.size()

**Review by robertapplin — APPROVED** [review]
Looks good to me 👍 Test failure is unrelated

**Pull request merged**